internal/mountinfo/mountinfo.go

```diff
@@ -0,0 +1,206 @@
+// contains the names of the block storage devices that back each of the requested mounts.
+//
+// Mounts is a set of name -> file path mappings (example: {"indexDir": "/home/.zoekt"}).
+//
+// sysMountPoint is the custom mount point for the sysfs pseudo-filesystem to use. If empty,
+// DefaultSysMountPoint will be used instead.
+//
+// The metric "mount_point_info" has a constant value of 1 and two labels:
+//   - mount_name: caller-provided name for the given mount (example: "indexDir")
+//   - device: name of the block device that backs the given mount file path (example: "sdb")
+//
+// This metric only works on Linux-based operating systems that have access to the sysfs pseudo-filesystem.
+// On all other operating systems, this metric will not emit any values.
+func MustRegisterNewMountPointInfoMetric(logger sglog.Logger, sysMountPoint string, mounts map[string]string) {
+	if sysMountPoint == "" {
+		sysMountPoint = DefaultSysMountPoint
+	}
+
+	metric := promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "mount_point_info",
+		Help: "An info metric with a constant '1' value that contains mount_name, device mappings",
+	}, []string{"mount_name", "device"})
+
+	// This device discovery logic relies on the sysfs pseudo-filesystem, which only exists
+	// on linux.
+	//
+	// See https://en.wikipedia.org/wiki/Sysfs for more information.
+	if runtime.GOOS != "linux" {
+		return
+	}
+
+	for mountName, mountFilePath := range mounts {
+		// For each <mountName>:<mountFilePath> pairing, we need to
+		// discover the name of the block device that stores <mountFilePath>.
+		//
+		// Example: For a mount named "indexDir" with a filePath of "~/.zoekt", this logic will
+		// discover that "sdb" is the disk that backs "~/.zoekt".
+		//
+		// Note: It's quite involved to implement the device discovery logic for
+		// every possible kind of storage device (e.x. logical volumes, NFS, etc.) See
+		// https://unix.stackexchange.com/a/11312 for more information.
+		//
+		//  As a result, this logic will only work correctly for mountFilePaths that are either:
+		// 	- stored directly on a block device
+		//	- stored on a block device's partition
+		//
+		// For all other device types, this logic will either:
+		// - find an incorrect device name
+		// - emit an error in the logs
+		//
+		// This logic was implemented from information gathered from the following sources (amongst others):
+		// - "The Linux Programming Interface" by Michael Kerrisk: Chapter 14
+		// - "Linux Kernel Development" by Robert Love: Chapters 13, 17
+		// - https://man7.org/linux/man-pages/man5/sysfs.5.html
+		// - https://en.wikipedia.org/wiki/Sysfs
+		// - https://unix.stackexchange.com/a/11312
+
+		// 'stat' the mount's mountFilePath path to determine what the device's ID numbers is
+		discoveryLogger := logger.Scoped("deviceNameDiscovery", "").
+			With(sglog.String("mountName", mountName)).
+			With(sglog.String("mountFilePath", mountFilePath))
```

Comment:
nit: you only need to call With once
```suggestion
		discoveryLogger := logger.Scoped("deviceNameDiscovery", "").With(
			sglog.String("mountName", mountName),
			sglog.String("mountFilePath", mountFilePath),
		)
```